File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] arr) {
        int n=arr.length;
        ArrayList<ArrayList<Integer>> adj=new ArrayList<>();
        for(int i=0;i<n;i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<arr[0].length;j++){
                if(i!=j && arr[i][j]==1){
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }
        int count=0;
        int visited[]=new int[n];
        for(int i=0;i<n;i++){
            if(visited[i]==0){
                count++;
                 dfs(i,adj,visited);  
            }
        }
        return count;
    }
    static void dfs(int node,ArrayList<ArrayList<Integer>> adj,int visited[]){
        visited[node]=1;
        for(int n:adj.get(node)){
            if(visited[n]==0){
                dfs(n,adj,visited);
            }
        }
    }
}

File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(int curr,vector<int>& visited,vector<vector<int>>& adj){
        visited[curr]=1;
        for(int nei:adj[curr]){
            if(visited[nei]==0){
                dfs(nei,visited,adj);
            }
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n=isConnected.size();
        vector<vector<int>> adj(n);
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(isConnected[i][j]==1 && i!=j){
                    adj[i].push_back(j);
                }
            }
        }
        int count=0;
        vector<int> visited(n,0);
        for(int i=0;i<n;i++){
            if(visited[i]==0){
                count++;
                dfs(i,visited,adj);
                
            }
        }
        return count;
    }
};